epstein/src/main/scala/epstein/ZQuery.scala

```diff
@@ -0,0 +1,211 @@
+package epstein
+
+import zio.{ Ref, ZIO }
+
+/**
+ * A `ZQuery[R, E, A]` is a purely functional description of an effectual query
+ * that may contain requests from one or more data sources, requires an
+ * environment `R`, may fail with an `E`, and may succeed with an `A`. All
+ * requests that do not need to be performed sequentially, as expressed by
+ * `flatMap` or combinators derived from it, will automatically be batched,
+ * allowing for aggressive data source specific optimizations. Requests will
+ * also automatically be deduplicated and cached.
+ *
+ * This allows for writing queries in a high level, compositional style, with
+ * confidence that they will automatically be optimized. For example, consider
+ * the following query from a user service.
+ *
+ * {{{
+ * val getAllUserIds: ZQuery[Any, Nothing, List[Int]] = ???
+ * def getUserNameById(id: Int): ZQuery[Any, Nothing, String] = ???
+ *
+ * for {
+ *   userIds   <- getAllUserIds
+ *   userNames <- ZQuery.foreachPar(userIds)(getUserNameById)
+ * } yield userNames
+ * }}}
+ *
+ * This would normally require N + 1 queries, one for `getAllUserIds` and one
+ * for each call to `getUserNameById`. In contrast, `ZQuery` will automatically
+ * optimize this to two queries, one for `userIds` and one for `userNames`,
+ * assuming an implementation of the user service that supports batching.
+ *
+ * Based on "There is no Fork: an Abstraction for Efficient, Concurrent, and
+ * Concise Data Access" by Simon Marlow, Louis Brandy, Jonathan Coens, and Jon
+ * Purdy. [[http://simonmar.github.io/bib/papers/haxl-icfp14.pdf]]
+ */
+sealed trait ZQuery[-R, +E, +A] { self =>
+
+  /**
+   * Executes one step of this query.
+   */
+  protected def step(cache: Cache): ZIO[R, E, Result[R, E, A]]
+
+  /**
+   * A symbolic alias for `zipPar`.
+   */
+  final def <&>[R1 <: R, E1 >: E, B](that: ZQuery[R1, E1, B]): ZQuery[R1, E1, (A, B)] =
+    zipPar(that)
+
+  /**
+   * Returns a query that models execution of this query, followed by passing
+   * its result to the specified function that returns a query. Requests
+   * composed with `flatMap` or combinators derived from it will be executed
+   * sequentially and will not be batched, though deduplication and caching of
+   * requests will still be applied.
+   */
+  final def flatMap[R1 <: R, E1 >: E, B](f: A => ZQuery[R1, E1, B]): ZQuery[R1, E1, B] =
+    new ZQuery[R1, E1, B] {
+      def step(cache: Cache): ZIO[R1, E1, Result[R1, E1, B]] =
+        self.step(cache).flatMap {
+          case Result.Blocked(br, c) => ZIO.succeed(Result.blocked(br, c.flatMap(f)))
+          case Result.Done(a)        => f(a).step(cache)
+        }
+    }
+
+  /**
+   * Maps the specified function over the result of this query.
+   */
+  final def map[B](f: A => B): ZQuery[R, E, B] =
```

Comment:
If `map` is implemented using `flatMap`, it will prevent mapped `ZQuery`s from being batched, no? How about implementing it directly?

Reply:
I don't think it was an issue because batching only applies to `ZQuery` instances and `map` returns an `A` rather than a `ZQuery` of `A`. But for a common method there is also no need to wrap it in an additional effect constructor and then have to unwrap it with `flatMap` so I implemented it directly.